### PR TITLE
Add ability to parameterize the save operator

### PR DESCRIPTION
**Pull request by kenxu95**
@likawind and @hsubbaraj-spiral as lead reviewers. Tagging @eunice-chan since you're more familiar with the `list_saved_objects()` part. 
@vsreekanti to watch the demo.

## Describe your changes and why you are making these changes
From an implementation perspective, for parameterized saves, we set the table name on the spec to a empty string if it's parameterized. This table name is then filled in on the spec at runtime, based on the input parameter, in the python executor.

The write part of this is actually quite easy. The tough part was getting this to work with read part of it, which is necessary for `delete_workflow()` and `list_saved_objects()`. This is because both these code paths rely on the LoadSpec in the database, which is empty. This means we have to do a ton of custom logic to go a find the input parameter artifact to these parameterized saves, and find the history of all parameter values of that artifact. See `GetDistinctSaveOpsByWorkflow()` for this logic.

## Related issue number (if any)
ENG-2675

## Loom demo (if any)
https://www.loom.com/share/1f96a44998a042f39bd344aad6306810

## Checklist before requesting a review
- [ ] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [ ] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [ ] If this is a new feature, I have added unit tests and integration tests.
- [ ] I have run the integration tests locally and they are passing.
- [ ] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

File: src/golang/cmd/server/handler/delete_workflow.go
Comment (kenxu95) [inline src/golang/cmd/server/handler/delete_workflow.go:211]: There may be an argument for using `GetDistinctSaveOpsByWorkflow` from the start, instead of going through `h.OperatorRepo.GetLoadOpsByWorkflowAndIntegration()` and then falling back to this method, but I don't understand the difference super well and didn't want to rock the boat too much. cc: @eunice-chan

Comment (eunice-chan) [reply src/golang/cmd/server/handler/delete_workflow.go:211]: I think you should be able to use `GetDistinctSaveOpsByWorkflow` from the get-go. For consistency's sake it might be be good to call save ops, load ops.

Comment (kenxu95) [reply src/golang/cmd/server/handler/delete_workflow.go]: Updated. Would appreciate another look! In particular I added a new `NonRelational` cast.

**Comment by vsreekanti** [comment]
The demo looks good to me! One thing we might want to do in follow-on work is allow string interpolation in the save operator like we do in the load operator, so I can do something like `table_{{param}}`. Can we create a followup task to track that?

**Review by likawind — APPROVED** [review]
Approving to unblock. Based on the demo, we could run another test using regular save op to make sure we don't have any regression. Two other comments that I don't think are blocking this PR:
* What does it look like when you click the operator node? Does it show the parameter as a input in `inputs` section?
* It appears we don't do string replacement like how we use in extract operators. I think it's fine as an early feature and if it fit's Yext's current need. I'm wondering if we should have a plan making it consistent with extracts. It would enable use cases like `save(table="my_results_{{ today }}")` .

File: src/golang/cmd/server/handler/list_workflow_objects.go
Comment (hsubbaraj-spiral) [inline src/golang/cmd/server/handler/list_workflow_objects.go:159]: How do we know the parameter artifact is the first input to the save operator?

**Review by hsubbaraj-spiral — APPROVED** [review]
The logic makes sense to me. Thanks for extensively commenting on GetDistinctLoadOpsByWorkflow, it helped clarify my questions. Left one question as a comment. Would also like to see some testing (unit, backend integration, or sdk integration test) but I understand that may not happen with this PR given the time constraint. Maybe we can come back to that in a later PR.

**Comment by kenxu95** [comment]
wrt to the string interpolation requests, the task is here: https://linear.app/aqueducthq/issue/ENG-2982/add-documentation-for-parameterized-sql-operators Note that we don't really use the `{{...}}` syntax anymore for interpolating extracts.

> What does it look like when you click the operator node? Does it show the parameter as input in inputs section?
Yeah I believe so.

> Would also like to see some testing (unit, backend integration, or sdk integration test) but I understand that may not happen with this PR given the time constraint.
There should be a fairly extensive integration test in this PR.

File: src/golang/cmd/server/handler/list_workflow_objects.go
Comment (kenxu95) [reply src/golang/cmd/server/handler/list_workflow_objects.go:159]: That's just always the order that the save() on the SDK adds it in. I'll add a comment in both places clarifying this assumption.

**Pull request merged**
